src/nvim/eval/funcs.c

```diff
@@ -2663,14 +2663,16 @@ static void f_fnamemodify(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   char buf[NUMBUFLEN];
   const char *fname = tv_get_string_chk(&argvars[0]);
   const char *const mods = tv_get_string_buf_chk(&argvars[1], buf);
-  if (fname == NULL) {
+  if (mods == NULL || fname == NULL) {
     fname = NULL;
-  } else if (mods != NULL && *mods != NUL) {
+  } else {
     len = strlen(fname);
-    size_t usedlen = 0;
     if (*mods != NUL) {
-      (void)modify_fname((char_u *)mods, false, &usedlen,
-                         (char_u **)&fname, &fbuf, &len);
+      size_t usedlen = 0;
+      if (*mods != NUL) {
+        (void)modify_fname((char_u *)mods, false, &usedlen,
+                           (char_u **)&fname, &fbuf, &len);
+      }
```

Comment:
This `*mods != NUL` already been checked
```suggestion
      (void)modify_fname((char_u *)mods, false, &usedlen,
                         (char_u **)&fname, &fbuf, &len);
```